### ISSUE TITLE
URGENT: 0.13.4 crashing after audio plays on iOS

### ISSUE THREAD

**Issue by lukasgit** [luanpotter/audioplayers#345]
- 0.13.3 had random crash on iOS app start
- 0.13.4 fixed random crash on iOS app start (yay!)
- 0.13.4 crash after audio plays on iOS (boo!)

I hope this issue can be resolved asap. Thanks!!

**Comment by lukasgit** [comment]
@subhash279 @luanpotter @erickzanardo @Afsar-Pasha

**Comment by imaNNeoFighT** [comment]
There is a bug happening in `0.13.4`.
After playing audio ios crashes without any console log (because crash happens in the native side),
here I got something in my XCode but I don't know anything about it.

![image](https://user-images.githubusercontent.com/7009300/70186305-e3f2cb00-1700-11ea-86cd-035f2ca164a7.png)


I've checked this [commit](https://github.com/luanpotter/audioplayers/commit/813006bf8250dac227814a88e0418f2ac5d20bb2)
```dart
    // Start the headless audio service. The parameter here is a handle to	
    // a callback managed by the Flutter engine, which allows for us to pass	
    // references to our callbacks between isolates.	
    final CallbackHandle handle =	
        PluginUtilities.getCallbackHandle(_backgroundCallbackDispatcher);	
    assert(handle != null, 'Unable to lookup callback.');	
    _invokeMethod('startHeadlessService', {	
      'handleKey': <dynamic>[handle.toRawHandle()]	
    });
```

these codes removed and wrapped in the `startHeadlessService` function, but it didn't call anywhere.

**Comment by jaymacdonald67** [comment]
This is happening to me as well. It just started yesterday. Before that, it was working well. Now it crashes as soon as the sound finishes playing. 

`flutter: _platformCallHandler call audio.onComplete {playerId: f0fd155b-63e7-4cad-b22e-06a2a554fe99}
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff23b98bde __exceptionPreprocess + 350
	1   libobjc.A.dylib                     0x00007fff503b5b20 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff23b98958 +[NSException raise:format:arguments:] + 88
	3   Foundation                          0x00007fff255eb6f5 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   Flutter                             0x000000010b0b9a19 -[FlutterEngine sendOnChannel:message:binaryReply:] + 607
	5   audioplayers                        0x000000010d195b67 -[AudioplayersPlugin onSoundComplete:] + 903
	6   audioplayers                        0x000000010d194267 __69-[AudioplayersPlugin setUrl:isLocal:isNotification:playerId:onReady:]_block_invoke.352 + 87
	7   Foundation                  <…>
Lost connection to device.`

**Comment by erickzanardo** [comment]
I have debugged and this line seems to be the problem https://github.com/luanpotter/audioplayers/blob/b3cbad8da5448471c3d9532c1d14329f8fbaa68e/ios/Classes/AudioplayersPlugin.m#L631, by removing it, the crash stops, but I have no idea what is the problem, I am not an experienced iOS developer, @subhash279 this seems to be related to your recently changes, if I am not wrong, do you have any idea?

**Comment by JesseScott** [comment]
I'm looking into this and can't determine how those changes have affected this - i see all the same channels and callbacks...

Fair disclosure, it's been a few years since i've dug into Objective-C though 😅

**Comment by JesseScott** [comment]
Okay, so... it was in the comments from @subhash279  📖 

```
  /// this should be called after initiating AudioPlayer only if you want to
  /// listen for notification changes in the background
  void startHeadlessService() { ... }
```

You simply have to call `startHeadlessService()` on the `AudioPlayer` (or `myPlayer.fixedPlayer` if you are using `AudioCache`)...

That being said, this should have been added to the example, and there should be some type of flag in `audioplayers/ios/Classes/AudioplayersPlugin.m` to avoid invoking this channel if this method isn't called... will see if I can make that happen.

**Comment by luanpotter** [comment]
I agree that we should have the check, one should not be asked to call `startHeadlessService` if one doesn't want to use this feature.

**Comment by JesseScott** [comment]
@luanpotter i've never developed a Flutter plugin before, is there a quickstart guide to running locally? I assume that i'd need to point my `pubspec.yaml` to a local gradle instance, or... ?

**Comment by luanpotter** [comment]
@JesseScott you can cd into the `example` folder and run the app there. That example app already links to your local version of the plugin. Basically you can use a relative path. The example app already has several options to make testing easy :)

**Comment by JesseScott** [comment]
ah of course!

**Comment by luanpotter** [comment]
To be more clear, I mean:

```bash
cd example
flutter run
```

or take a look at the example package `pubspec.yaml` how to link your other project locally :)

**Comment by JesseScott** [comment]
yep, got it

**Comment by JesseScott** [comment]
Hmmm. 

I'm a little lost on how to send a flag into the iOS portion of the lib, or have that portion check on anything from the Dart side - not too familiar with the isolate channel invocation... 

Open to thoughts on how i'd properly solve this...

**Comment by JesseScott** [comment]
actually, think i have something...

**Comment by JesseScott** [comment]
PR is up @luanpotter @erickzanardo @subhash279

**Comment by luanpotter** [comment]
merged and released in 0.13.5 :) thanks for the help, @JesseScott

**Comment by lukasgit** [comment]
@JesseScott @luanpotter thank you SO VERY MUCH!!!
